mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java

```diff
@@ -70,138 +68,116 @@ public ReIndexerPerformer(MailboxManager mailboxManager,
-        }
+                .switchIfEmpty(Mono.just(Result.COMPLETED))
+                .doFinally(any -> LOGGER.info("Finish to reindex mailbox with mailboxId {}", mailbox.getMailboxId().serialize())));
     }
 
-    Task.Result reIndex(ReprocessingContext reprocessingContext, ReIndexingExecutionFailures previousReIndexingFailures) {
-        return previousReIndexingFailures.failures()
-            .stream()
-            .map(previousFailure -> reIndex(reprocessingContext, previousFailure))
+    Mono<Result> reIndex(ReprocessingContext reprocessingContext, ReIndexingExecutionFailures previousReIndexingFailures) {
+        return Flux.fromIterable(previousReIndexingFailures.failures())
+            .flatMap(previousFailure -> reIndex(reprocessingContext, previousFailure), MESSAGE_CONCURRENCY)
             .reduce(Task::combine)
-            .orElse(Task.Result.COMPLETED);
+            .switchIfEmpty(Mono.just(Result.COMPLETED));
     }
 
-    private Task.Result reIndex(ReprocessingContext reprocessingContext, ReIndexingExecutionFailures.ReIndexingFailure previousReIndexingFailure) {
+    private Mono<Result> reIndex(ReprocessingContext reprocessingContext, ReIndexingExecutionFailures.ReIndexingFailure previousReIndexingFailure) {
         MailboxId mailboxId = previousReIndexingFailure.getMailboxId();
         MessageUid uid = previousReIndexingFailure.getUid();
-        try {
-            return handleMessageReIndexing(mailboxId, uid, reprocessingContext);
-        } catch (MailboxException e) {
-            LOGGER.warn("ReIndexing failed for {} {}", mailboxId, uid, e);
-            reprocessingContext.recordFailureDetailsForMessage(mailboxId, uid);
-            return Task.Result.PARTIAL;
-        }
+
+        return handleMessageReIndexing(mailboxId, uid, reprocessingContext)
+            .onErrorResume(e -> {
+                LOGGER.warn("ReIndexing failed for {} {}", mailboxId, uid, e);
+                reprocessingContext.recordFailureDetailsForMessage(mailboxId, uid);
+                return Mono.just(Result.PARTIAL);
+            });
     }
 
-    Task.Result reIndex(ReprocessingContext reprocessingContext) throws MailboxException {
+    Mono<Result> reIndex(ReprocessingContext reprocessingContext) {
         MailboxSession mailboxSession = mailboxManager.createSystemSession(RE_INDEXER_PERFORMER_USER);
         LOGGER.info("Starting a full reindex");
-        Stream<MailboxId> mailboxIds = mailboxSessionMapperFactory.getMailboxMapper(mailboxSession).list()
-            .stream()
-            .map(Mailbox::getMailboxId);
-
-        try {
-            return reIndex(mailboxIds, reprocessingContext);
-        } finally {
-            LOGGER.info("Full reindex finished");
-        }
+        return mailboxSessionMapperFactory.getMailboxMapper(mailboxSession).list()
+            .concatMap(mailbox -> reIndex(reprocessingContext, mailboxSession, mailbox))
```

Comment:
you don't care about ordering here, don't you?

Reply:
This is a strategy for concurrency control until we get better.

I intend toi start discussions on this, but I see reactification as a preliminary step.

Reply:
So: 

 1. Can we merge the reactified version of this as is?

Regarding concurrency control, I propose to let the admin specify the concurrent count of mailboxes & messages he is wishing to support concurrently.

We could do that: 

   - As part of task definition: each task defines its concurrency parameters. 

Pros: easy to implement and incremental
Cons: each task need to be adapted, task serialization needs to be adapted

   - Add a configuration option on the TaskManager: "task name -> configuration option -> value". Then each task, when running could be readingthis configuration. 

Pros: easy to implement, no serialization/webadmin handling
Cons: each task needs to be modified, reboot needed to update that configuration

IMO a no go.

 - Add a generic `RunningContext` parameter that is passed upon `TaskManager::submit`. It could be retrieved by webadmin from query parameters and handled in a generic way. Then we store it along with tasks. When executed tasks use it to adjust their concurrency (`Task::run` takes a RunningContext parameter).

Pros: Generic webadmin layer
Cons: All tasks needs to handle (or not) this generic parameter. Serialization might need to be adapted. Concurrency control is not that generic. It could be multi-leyer: ex user -> mailbox -> message... Fitting any specific examples into a unified generic model could be hazardous.

WDYT?

Reply:
I definitely prefer the first solution: concurrency will probably be express in the best unit depending on the task so it make sense to have it in the task configuration.

Reply:
Let's then do that in a follow up PR if you don't mind :-)

Reply:
I approved the PR yesterday BTW

Reply:
Your approuval was 22 hour ago, my comment 23 hour ago, so at the time of writting it, insisting on the follow up strategy was relevant ;-)